typescript/sdk/src/fee/EvmTokenFeeModule.ts

```diff
@@ -141,23 +181,117 @@ export class EvmTokenFeeModule extends HyperlaneModule<
     return deployer.deploy({ [params.chainName]: params.config });
   }
 
-  // Public accessor for the deployed fee contract address
-  getDeployedFeeAddress(): Address {
-    return this.args.addresses.deployedFee;
-  }
+  async read(params?: Partial<ReaderParams>): Promise<DerivedTokenFeeConfig> {
+    const address = params?.address ?? this.args.addresses.deployedFee;
+    const routingDestinations = params?.routingDestinations ?? [];
 
-  async read(routingDestinations?: number[]): Promise<DerivedTokenFeeConfig> {
-    return this.reader.deriveTokenFeeConfig(
-      this.args.addresses.deployedFee,
+    return this.reader.deriveTokenFeeConfig({
+      address: address,
       routingDestinations,
-    );
+    });
   }
 
   async update(
-    _config: TokenFeeConfigInput,
-  ): Promise<
-    Annotated<ProtocolTypedTransaction<ProtocolType.Ethereum>['transaction']>[]
-  > {
-    throw new Error('Not implemented');
+    targetConfig: TokenFeeConfigInput,
+    params?: Partial<ReaderParams>,
+  ): Promise<AnnotatedEV5Transaction[]> {
+    let updateTransactions: AnnotatedEV5Transaction[] = [];
+
+    const normalizedTargetConfig: TokenFeeConfig = normalizeConfig(
+      await EvmTokenFeeModule.expandConfig({
+        config: targetConfig,
+        multiProvider: this.multiProvider,
+        chainName: this.chainName,
+      }),
+    );
+
+    const actualConfig = await this.read(params);
+    const normalizedActualConfig: TokenFeeConfig =
+      normalizeConfig(actualConfig);
+
+    //If configs are the same, return empty array
+    if (deepEquals(normalizedActualConfig, normalizedTargetConfig)) {
+      this.logger.debug(
+        `Same config for ${normalizedTargetConfig.type}, no update needed`,
+      );
+      return [];
+    }
+
+    // if the type is a immutable type, then redeploy
+    if (
+      ImmutableTokenFeeType.includes(
+        normalizedTargetConfig.type as (typeof ImmutableTokenFeeType)[number],
+      )
+    ) {
+      this.logger.info(
+        `Immutable fee type ${normalizedTargetConfig.type}, redeploying`,
+      );
+      const contracts = await this.deploy({
+        config: normalizedTargetConfig,
+        multiProvider: this.multiProvider,
+        chainName: this.chainName,
+        contractVerifier: this.contractVerifier,
+      });
+      this.args.addresses.deployedFee =
+        contracts[this.chainName][normalizedTargetConfig.type].address;
+    }
+
+    // if the type is a mutable type, then update
+    updateTransactions = [
+      ...(await this.updateRoutingFee(
+        _.merge(
+          actualConfig,
+          normalizedTargetConfig,
+        ) as DerivedRoutingFeeConfig,
+      )),
+      ...this.createOwnershipUpdateTxs(
+        normalizedActualConfig,
+        normalizedTargetConfig,
+      ),
+    ];
+
+    return updateTransactions;
+  }
+
+  private async updateRoutingFee(targetConfig: DerivedRoutingFeeConfig) {
```

Comment:
Shouldn't we also check if there are fee contracts that should be removed from the config? is it possible to unset a fee contract for a given domain?

Reply:
ah good catch.

Yes, its the same method, but with `feeContract` as address(0)